ext/liquid_c/block.c

```diff
@@ -79,28 +68,45 @@ static VALUE rb_block_parse(VALUE self, VALUE tokens, VALUE options)
                 rb_ary_push(nodelist, str);
 
                 if (rb_ivar_get(self, intern_blank) == Qtrue) {
-                    const char *end = token.str + token.length;
+                    const char *end = token.str_full + token.len_full;
 
-                    if (read_while(token.str, end, rb_isspace) < end)
+                    if (read_while(token.str_full, end, rb_isspace) < end)
                         rb_ivar_set(self, intern_blank, Qfalse);
                 }
                 break;
             }
             case TOKEN_VARIABLE:
             {
-                VALUE args[2] = {rb_enc_str_new(token.str + 2 + token.lstrip, token.length - 4 - token.lstrip - token.rstrip, utf8_encoding), options};
+                VALUE args[2] = {rb_enc_str_new(token.str_trimmed, token.len_trimmed, utf8_encoding), parse_context};
                 VALUE var = rb_class_new_instance(2, args, cLiquidVariable);
                 rb_ary_push(nodelist, var);
                 rb_ivar_set(self, intern_blank, Qfalse);
                 break;
             }
             case TOKEN_TAG:
             {
-                const char *start = token.str + 2 + token.lstrip, *end = token.str + token.length - 2 - token.rstrip;
+                const char *start = token.str_trimmed, *end = token.str_trimmed + token.len_trimmed;
 
                 // Imitate \s*(\w+)\s*(.*)? regex
                 const char *name_start = read_while(start, end, rb_isspace);
                 const char *name_end = read_while(name_start, end, is_id);
+                long name_len = name_end - name_start;
+
+                if (name_len == 0) {
+                    VALUE str = rb_enc_str_new(token.str_trimmed, token.len_trimmed, utf8_encoding);
+                    return rb_yield_values(2, str, str);
+                }
+
+                if (strncmp(name_start, "liquid", name_len < 6 ? name_len : 6) == 0) {
+                    const char *liquid_tag_start = read_while(name_end, end, rb_isspace);
+                    int line_number = token_start_line_number;
+                    if (line_number) {
+                        line_number += count_newlines(token.str_full, liquid_tag_start);
+                    }
+                    VALUE liquid_tag_tokenizer = tokenizer_new_from_cstr(liquid_tag_start, end, line_number, true);
```

Comment:
It might make more sense to refactor the tokenizer so that it keeps track of this liquid tag state and avoids the need for the block to know anything about it.  This seems like it would be harder to do with the liquid regex approach, but seems like it would be easier to do in C.  This could also be used to remove the whitespace trimming concern from the block body in liquid-c as well.

For instance, I see there being the following states, which could be represented by an enum for a switch statement or using functions to store on a function pointer:
* TOKENIZE_TEMPLATE
* TOKENIZE_TAG
* TOKENIZE_VARIABLE
* TOKENIZE_LIQUID_TAG

Even now, at the end of parsing a TOKEN_RAW type token, we already know whether the next token should be a TOKEN_TAG or TOKEN_VARIABLE.  So refactoring the current code would allow us to resume parsing without losing this known state.

You are essentially already keeping track of some parse state already, but right now that is being done as a `for_liquid_tag` boolean state.  I don't think we even need to support the `for_liquid_tag` option to Liquid::Tokenizer.new.

Reply:
This refactor could be done later.  My review comments below should be much faster to address, such that we can get this merged and point the liquid repo back to the master branch of liquid-c.